virtcontainers/pkg/types/pcipath.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2020 David Gibson, Red Hat Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package types
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+const maxPciSlot = 31
+
+// A PciPath describes where a PCI sits in a PCI hierarchy.
+//
+// Consists of a list of PCI slot numbers, giving the slot of each
+// bridge that must be traversed from the PCI root to reach the
+// device, followed by the slot of the device itself
+//
+// When formatted into a string is written as "xx/.../yy/zz" Here, zz
+// is the slot of the device on its PCI bridge, yy is the slot of the
+// bridge on its parent bridge and so forth until xx is the slot of
+// the "most upstream" bridge on the root bus.  If a device is
+// connected directly to the root bus, its PciPath is just "zz"
+
+type PciPath struct {
+	path []uint8
+}
+
+func (p PciPath) Format() string {
+	tokens := make([]string, len(p.path))
+	for i, slot := range p.path {
+		tokens[i] = fmt.Sprintf("%02x", slot)
+	}
+	return strings.Join(tokens, "/")
+}
+
+func (p PciPath) IsNil() bool {
+	return p.path == nil
+}
+
+func PciPathFromString(s string) (PciPath, error) {
+	if s == "" {
+		return PciPath{}, nil
+	}
+
+	tokens := strings.Split(s, "/")
+	slots := make([]uint8, len(tokens))
+	for i, t := range tokens {
+		slot, err := strconv.ParseUint(t, 16, 5)
```

Comment:
This seems trivial but there are three `strconv.ParseUint()` calls. I wonder if we should create a `strToSlot(s string) (uint8, error)` function to encapsulate this? As a bonus that function to also check the value against `maxPciSlot`.

Reply:
I can't really see a sensible way to do this, because 2 of the `ParseUint` calls are outside the code specifically for the new data structure.  `MakePciPath()` still needs to validate against `maxPciSlot`, so checking in a helper before passing into there would be redundant.

This is kind of unavoidably ugly, because those callers have one of their path components as a string, and another as a number.  I'd like to change that, but that involves refactoring of a bunch of other things that's not really in scope for this PR.

Reply:
I think I found a way which is done in the new push.